### ISSUE TITLE
Forwarding "login page" breaks requests to hosted APIs (missing CORS headers, "signing in" HTML page)

### ISSUE THREAD

**Issue by DanTup** [MicrosoftDocs/vsonline#164]
This might be covered by https://github.com/MicrosoftDocs/vsonline/issues/19, but I'm not sure. My requests are originating from a Chrome extension and not directly in a page being served by VS Online.

When the Chrome extension sends the first request to the server, it gets served up a page like:

```
Visual Studio Online
You need to sign in first, redirecting...

Click here if your browser does not automatically redirect.
```

As well as not being the service that the extension expected, it does not contain the CORS headers that the real service serves up, so the Chrome extension fails.

The service was exposed using VS Code's `asExternalUrl` API.

**Comment by DanTup** [comment]
Below is the full contents of the page that's being served up. I guess it's trying to redirect through a page that sets an auth cookie or something? I don't know if it's related, but I'm guessing the service worker won't work here (the request originates in a Chrome extension), so this may be more complicated than just improving the in-browser forwarding.

```
<!DOCTYPE html>
<html>
<head>
    <environment include="Development">
        <link rel="stylesheet" href="/site.css" />
    </environment>
</head>
<body>
    <div class="vsonline-titlebar part">
        <div class="vsonline-titlebar__caption" aria-label="Visual Studio logo">
            <div class="vsonline-titlebar__logo"></div>
            <div class="vsonline-titlebar__separator"></div>
            <div class="vsonline-titlebar__caption-text">
                Visual Studio Online
            </div>
        </div>
    </div>
    <div>
        
<!doctype html>
<html lang="en">
<head>
    <title>Port Forwarding exceptions...</title>
</head>
<body>
    <div class="vsonline__redirect">
                <p>You need to sign in first, redirecting...</p>
                <p>Click <a href="https://online.visualstudio.com/login?redirectUrl=https%3a%2f%2fb6eba251d5e2a55cbb70b13de2379c8264AA-43171.app.online.visualstudio.com%2f%24debug">here</a> if your browser does not automatically redirect.</p>
                <div class="vsonline__spinner"></div>
                <script type="text/javaScript">
                    setTimeout(function() {
                        window.location = "https://online.visualstudio.com/login?redirectUrl=https%3a%2f%2fb6eba251d5e2a55cbb70b13de2379c8264AA-43171.app.online.visualstudio.com%2f%24debug";
                    }, 5000);
                </script>
    </div>
</body>
</html>
    </div>
</body>
</html>
```

**Comment by treeder** [comment]
This issue seems seems to be the cause of a CORS error preventing making it so you can't test a web front end from your browser that's accessing an API hosted on a codespace. You'll get this error:

```
Access to fetch at 'https://ABC-8080.apps.codespaces.githubusercontent.com/orgs' from origin 'https://XYZ-5000.apps.codespaces.githubusercontent.com' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: Redirect is not allowed for a preflight request.
```

**Comment by ruohki** [comment]
having the same issue a apollo graphqlserver that is hosting graphql playground in /graphql cant be accessed due to cors


```
Access to fetch at 'https://github.com/codespaces/auth/8fdfb7ff-5131-4ec5-802f-5a91f9eaad20?path=%2fgraphql&port=4000&cid=e9f7d24533a8ad5050e30f45a809ec4e' (redirected from 'https://8fdfb7ff-5131-4ec5-802f-5a91f9eaad20-4000.apps.codespaces.githubusercontent.com/graphql') from origin 'https://8fdfb7ff-5131-4ec5-802f-5a91f9eaad20-4000.apps.codespaces.githubusercontent.com' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
```

**Comment by lostintangent** [comment]
@plisy Are we already tracking this issue? Seems like a potentially blocking problem for many scenarios?

**Comment by legomushroom** [comment]
That's the `fetch` limitation for the PFS domains, not the CORS issue. The fetch does not have the `credentials: include` option by default and since PortForwarding services relies on cookies for auth, the `fetch` request will be redirected and the browser will surface it as CORS issue since there is no `Access-Control-` headers in the response.

**Comment by abdulghanitech** [comment]
> That's the `fetch` limitation for the PFS domains, not the CORS issue. The fetch does not have the `credentials: include` option by default for the cross-origin requests, and since PortForwarding services relies on cookies for auth, the `fetch` request will be redirected and the browser will surface it as CORS issue since there is no `Access-Control-` headers in the response.
> 
> Can be fixed by having the public ports.

Any quick workaround?

**Comment by abdulghanitech** [comment]
> This issue seems seems to be the cause of a CORS error making it so you can't test a web front end from your browser that's accessing an API hosted on a codespace. You'll get this error:
> 
> ```
> Access to fetch at 'https://ABC-8080.apps.codespaces.githubusercontent.com/orgs' from origin 'https://XYZ-5000.apps.codespaces.githubusercontent.com' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: Redirect is not allowed for a preflight request.
> ```

How to solve though?

**Comment by legomushroom** [comment]
@Abdelrahmanpro I don't think we have a good quick workaround for that, besides the client should always include the credentials in the dev mode. We are planning to add the public endpoints for the forwarded ports soon that will not suffer from this issue anymore. Another idea we were thinking about is to add a bearer token and a custom header that you can use in the dev mode of your app.

For the second one, the app CPRS policies should also account for the `X-Forwarded-Host` header, so the CORS handshake will succeed. That is because we override the `Host` and `Origin` headers back to `localhost` so the `Access-Control-Allow-*` headers will be set to localhost by default in the CORS handshake flow.

cc @lostintangent @plisy

**Comment by darrenbutcher** [comment]
Came across the same issue as @abdulghanitech. Was doing some hair pulling, thought I missed something in my code, this is definitely a blocking issue. My current dev setup requires the use additional ports for other tools in dev mode that sync back with my application code. Pity there seems to be no workaround for this at the moment.
